src/viewer/viewer.ts

```diff
@@ -1,7 +1,113 @@
+        const viewport = page.getViewport({scale: 1})
+
+        const div = document.createElement("div")
+        div.id = `page-container-${i}`
+        div.style.width = `${Math.floor(viewport.width)}px`
+        div.style.height = `${Math.floor(viewport.height)}px`
+
+        pagesContainer.appendChild(div)
+        pages.push(div)
+    }
+    return pages
+}
+
+async function loadVisiblePages(pdf: PDFDocumentProxy,
+                                pages: HTMLElement[],
+                                pagesLoaded: Map<number, boolean>) {
+    // Binary search for first page to load
+    let left = 0
+    let right = pages.length - 1
+
+    while (left != right) {
+        const idx = Math.floor((left + right) / 2)
+        const bottom = pages[idx].getBoundingClientRect().bottom + window.innerHeight
+        if (bottom < 0) {
+            left = idx + 1
+        } else {
+            right = idx
+        }
+    }
+
+    // Linear search for last page to load
+    const pagesToLoad: Set<number> = new Set()
+    do {
+        pagesToLoad.add(left)
+    } while (++left < pages.length && pages[left].getBoundingClientRect().top <= 2 * window.innerHeight)
+
+    // Unload unneeded pages
+    for (const pageIdx of pagesLoaded.keys()) {
+        if (!pagesToLoad.has(pageIdx) && pagesLoaded.get(pageIdx) === true) {
+            pagesLoaded.delete(pageIdx)
+            pages[pageIdx].replaceChildren()
+        }
+    }
+
+    // Loop through unloaded pages
+    for (const pageIdx of pagesToLoad) {
+        if (!pagesLoaded.has(pageIdx)) {
+            const promise = pdf.getPage(pageIdx + 1).then(page => {
+                const viewport = page.getViewport({scale: 1})
+
+                const canvas = document.createElement("canvas")
+                canvas.width = Math.floor(viewport.width)
+                canvas.height = Math.floor(viewport.height)
+
+                const div = pages[pageIdx]
+                canvas.style.width = div.style.width
+                canvas.style.height = div.style.height
+
+                page.render({
+                    canvasContext: canvas.getContext("2d")!,
+                    viewport: viewport
+                })
+
+                div.appendChild(canvas)
+                pagesLoaded.set(pageIdx, true)
+            })
+
+            pagesLoaded.set(pageIdx, false)
+        }
+    }
+}
+
+// Handle invalid URLs and allow entering a new one
+function showUrlError(message: string) {
+    $("#url-error-message").innerText = message
+    $("#url-error-container").style.display = "block"
+
+    $("#url-error-input-submit").onclick = () => {
+        const newUrlInput = $<HTMLInputElement>("input#url-error-input")
+        document.location = `${document.location.origin}${document.location.pathname}?url=${newUrlInput.value}`
+    }
+}
+
+async function main() {
+    // Get URL
+    const urlParam = (new URL(document.location.href)).searchParams.get("url")
+
+    if (urlParam == null) {
```

Comment:
I believe the urlParam will be undefined as opposed to null so this won't fire when it should. I think you can just change this to` if(! urlParam)`  because undefined is falsy.

Reply:
You're right, not working because of string just checked. Good catch

Reply:
So much stuff is _Because Javascript_